Newsletters: add 223 (2022-10-26) (#902)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2022-10-26-newsletter.md

---
title: 'Bitcoin Optech Newsletter #223'
permalink: /en/newsletters/2022/10/26/
name: 2022-10-26-newsletter
slug: 2022-10-26-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME

## News

- **Continued discussion about full RBF:** in [last week's
  newsletter][news222 rbf], we summarized a discussion on the
  Bitcoin-Dev mailing list about the inclusion of a new `mempoolfullrbf`
  option that could create problems for a several businesses which

>   accept transactions with zero confirmations ("zeroconf") as final

Reviewer: nit: it says "zeroconf" here, but "zero conf" with a space is used for the rest of the document.

  payments.  Discussion continued this week on both the mailing list and
  the #bitcoin-core-dev IRC room.  Some highlights of the discussion
  include:

    - *Free option problem:* Sergej Kotliar [warned][kotliar free

>       option] that he believes the greatest problem with transaction
>       replacement is that it creates a free American call option.  For

Reviewer: My sincerest apologies if I seem argumentative here, but I interpreted his message differently. To me, it seemed that he was describing an existing problem that is worsened with easier replacements - not to dismiss this point, just that "create" doesn't seem like the right word.
```suggestion
      option] that he believes the greatest problem with transaction
      replacement is that it eases a free American call option.  For
```
![image](https://user-images.githubusercontent.com/25183001/197531516-dbdb984f-ecd2-4b58-a1e0-4eea702afc12.png)

Author: First, please argue!  Second, I phrased this section around "replacements" in general, rather than "full RBF" in particular, in order to try to convey the point that it's any type of replacement that *creates* the problem.  I'll edit to see if I can make that clearer.

Reviewer: Ah, I see I didn't catch that this was about replacements in general!

        transaction using [CPFP][topic cpfp], particularly if [package
        relay][topic package relay] was enabled.

        Antoine Riard [noted][riard free option] that the same risk
        exists with LN, as Alice could wait to pay merchant Bob's
        invoice up until shortly before it expired, giving her time to
        wait for the exchange rate to change.  Although in that case, if
        Bob noticed that the exchange rate had changed significantly, he
        could instruct his node not to accept the payment, returning the
        money to Alice.

    - *Bitcoin Core not in charge of network:* Gloria Zhao [wrote][zhao
      no control] in the IRC discussion, "I think whatever option we
      take, it should be made abundantly clear to users that Core does
      not control whether full RBF happens or not. We could revert
      [25353][bitcoin core #25353] and it could still happen. [...]"

    - *No removal means problem could happen:* in the IRC discussion,
      Anthony Towns [echoed][towns uncoordinated] his points from last
      week, "if we're not going to remove the mempoolfullrbf option from
      24.0, we're going for an uncoordinated deployment."

>         Greg Sanders was [doubtful][sanders doubt], "the question is
>         will 5%+ set a variable. I suspect not."  Towns [replied][towns

Reviewer 2: I'm not sure how far we want to edit direct quotes, but the “is will” throws me off a bit. Perhaps it could be smoothed out by adding a “[whether]” or comma, or this?

```suggestion
        Greg Sanders was [doubtful][sanders doubt], "the question is
        'will 5%+ set a variable?' I suspect not."  Towns [replied][towns
```

    carefully evaluate the risks, perhaps starting by reading the emails
    linked in [last week's newsletter][news222 rbf].

- **CoreDev.tech transcripts:** prior to The Atlanta Bitcoin Conference
  (TabConf), several developers participated in a CoreDev.tech event.
  [Transcripts][coredev xs] of several of the meetings from the event
  have been provided by Bryan Bishop.  Notable discussions included:

    - [Transport encryption][p2p encryption]: a conversation about the
      recent update to the [version 2 encrypted transport
      protocol][topic v2 p2p transport] proposal (see
      [Newsletter #222][news222 bip324]).  This protocol would make it
      harder for network eavesdroppers to learn which IP address
      originated a transaction and improve the ability to detect and
      resist man-in-the-middle attacks between honest nodes.

        The discussion covers several of the protocol design
        considerations and is a recommended read for anyone wondering
        why the protocol authors made certain decisions.  It also
        examines the relationship to the earlier [countersign][topic
        countersign] authentication protocol.

>     - [Fee market][]: a wide-ranging discussion about transaction fees

Reviewer 2: The good being bid on is blockspace, not fees. So isn't it a “blockspace market” rather than a fee market?

```suggestion
    - [Blockspace market][]: a wide-ranging discussion about transaction fees
```

      historically, presently, and in the future.  Some topics included
      queries about why blocks are seemingly almost always nearly full

>       but the mempool, debate about how long we have for a significant

Reviewer: "but the mempool," seems to be missing a few words maybe?

      fee market to develop before we have to [worry][topic fee sniping]
      about Bitcoin's long-term stability, and what solutions we could
      deploy if we did believe a problem existed.

    - [FROST][]: a presentation about the FROST threshold signature
      scheme.  The transcript documents several excellent technical

>       questions about the cryptographic choices in of the design and may

Reviewer: ```suggestion
      questions about the cryptographic choices in the design and may
```

      be interesting reading for anyone interested in learning more
      about FROST specifically or cryptographic protocol design in
      general.  See also the TabConf transcript about [ROAST][], another
      threshold signature scheme for Bitcoin.

    - [GitHub][github chat]: a discussion about moving the Bitcoin Core
      project's git hosting from GitHub to another issue and PR
      management solution, as well as considering the benefits of
      possibly continuing to use GitHub.

    - [Provable specifications in BIPs][hacspec chat]: a discussion
      about using the [hacspec][] specification language in BIPs to
      provide specifications which are provably correct.  See also the

>       [transcript][hacspec preso] for a related talk during the TabConf.

Reviewer 3: Unfortunately I only caught the tail-end of this one, and therefore the notes aren't very useful, and maybe shouldn't be included here. The tabconf transcript has the full talk.

      individual miners (those with local mining equipment) to choose
      which transactions to mine rather than the pool choosing
      transactions.  In addition to many other benefits, it was
      mentioned in the discussion that allowing individual miners to
      chose their own block template might become highly desirable to
      pools that are worried about governments mandating which
      transactions can be mined, as in the [Tornado Cash][] controversy.
      Most of the discussion focused on the changes that would need to
      be made to Bitcoin Core to enable native support for Stratum v2.
      See also the TabConf transcript about [Braidpool][braidpool chat],
      a decentralized pooled mining protocol.

    - [Merging][merging chat] is a discussion about strategies to help
      get code reviewed in the Bitcoin Core project, although many
      suggestions also apply to other projects.  Ideas included:

        - Break big changes into several small PRs

        - Make it easy for reviewers to understand the ultimate
          objective.  For all PRs, this means writing a motivational PR
          description.  For changes that are being made incrementally,

>           use tracking issues, project boards, and also motivate
>           refactorings by also opening the PRs that will use that

Reviewer 2: Two “also”

```suggestion
          use tracking issues, project boards, and motivate
          refactorings by also opening the PRs that will use that
```

## Selected Q&A from Bitcoin Stack Exchange

*[Bitcoin Stack Exchange][bitcoin.se] is one of the first places Optech
contributors look for answers to their questions---or when we have a
few spare moments to help curious or confused users.  In
this monthly feature, we highlight some of the top-voted questions and
answers posted since our last update.*

{% comment %}<!-- https://bitcoin.stackexchange.com/search?tab=votes&q=created%3a1m..%20is%3aanswer -->{% endcomment %}
{% assign bse = "https://bitcoin.stackexchange.com/a/" %}

FIXME:bitschmidty

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [Bitcoin Core 24.0 RC2][] is a release candidate for the
  next version of the network's most widely used full node
  implementation.  A [guide to testing][bcc testing] is available.

  **Warning:** this release candidate includes the `mempoolfullrbf`
  configuration option which several protocol and application developers
  believe could lead to problems for merchant services as described

>   in the newsletters for this week and last week.

Reviewer: Thanks for adding a note about mempoolfullrbf! I hope it causes people to test rc2 and consider their zeroconf setups. I also hope it doesn't cause people to think Bitcoin Core is deliberately/thoughtlessly putting anyone at risk.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo], [Core
Lightning][core lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
Proposals (BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*

> - [Bitcoin Core #23443][] p2p: Erlay support signaling FIXME:harding

Reviewer: Note: I suggest including a clarification (just a few words maybe) that transaction reconciliation functionality has *not* been added yet, only signaling/negotiation. Commenting because people seemed to be really confused by this on Twitter.

Users should *not* pull from master and turn this on thinking they're enabling Erlay on their node. If someone starts a node with this on right now and it's still running when Bitcoin Core releases software with Erlay fully enabled, their node might get disconnected because they're signaling support for Erlay but not actually doing things. (This goes for any other Erlay PRs merged before completion).

      same exchange rate risk.

        Greg Sanders and Jeremy Rubin [replied][sanders cpfp] in
        [separate][rubin cpfp] emails to note that merchant Bob could
        incentivize miners to confirm customer Alice's original
        transaction using [CPFP][topic cpfp], particularly if [package
        relay][topic package relay] was enabled.

        Antoine Riard [noted][riard free option] that the same risk
        exists with LN, as Alice could wait to pay merchant Bob's
        invoice up until shortly before it expired, giving her time to
        wait for the exchange rate to change.  Although in that case, if
        Bob noticed that the exchange rate had changed significantly, he
        could instruct his node not to accept the payment, returning the
        money to Alice.

    - *Bitcoin Core not in charge of network:* Gloria Zhao [wrote][zhao
      no control] in the IRC discussion, "I think whatever option we
      take, it should be made abundantly clear to users that Core does
      not control whether full RBF happens or not. We could revert
      [25353][bitcoin core #25353] and it could still happen. [...]"

>     - *No removal means problem could happen:* in the IRC discussion,

Author: Took the "the" change; left the case change (we usually lowercase after a colon).

      24.0, we're going for an uncoordinated deployment."

        Greg Sanders was [doubtful][sanders doubt], "the question is
        will 5%+ set a variable. I suspect not."  Towns [replied][towns
        uasf], "[UASF][topic soft fork activation] `uacomment`
        demonstrated it's easy to get ~11% to set a variable in just a
        couple of weeks".

    - *Should be an option:* Martin Zumsande [said][zumsande option] in
      the IRC discussion, "I think that if a meaningful number of node
      operators and miners want a specific policy, it shouldn't be on
      the devs to tell them 'you can't have that now'. Devs can and
      should give a recommendation (by picking the default), but
      providing options to informed users should never be a problem."

    As of this writing, no clear resolution had been reached.  The
    `mempoolfullrbf` option is still included in the release candidates
    for the upcoming version of Bitcoin Core 24.0 and it is Optech's
    recommendation that any service depending on zero conf transactions
    carefully evaluate the risks, perhaps starting by reading the emails
    linked in [last week's newsletter][news222 rbf].

> - **CoreDev.tech transcripts:** prior to The Atlanta Bitcoin Conference

Author: Leaving as-is.

  (TabConf), several developers participated in a CoreDev.tech event.

>   [Transcripts][coredev xs] of several of the meetings from the event

Reviewer 4: Couple items that might be useful for the reader regarding the event, include at your discretion:

- 40 developers
- transcripts only represent about half of the meetings that occurred
- meeting topics were chosen by attendees / unconference style

      queries about why blocks are seemingly almost always nearly full
      but the mempool, debate about how long we have for a significant
      fee market to develop before we have to [worry][topic fee sniping]
      about Bitcoin's long-term stability, and what solutions we could
      deploy if we did believe a problem existed.

    - [FROST][]: a presentation about the FROST threshold signature
      scheme.  The transcript documents several excellent technical
      questions about the cryptographic choices in of the design and may
      be interesting reading for anyone interested in learning more
      about FROST specifically or cryptographic protocol design in
      general.  See also the TabConf transcript about [ROAST][], another
      threshold signature scheme for Bitcoin.

    - [GitHub][github chat]: a discussion about moving the Bitcoin Core
      project's git hosting from GitHub to another issue and PR
      management solution, as well as considering the benefits of
      possibly continuing to use GitHub.

    - [Provable specifications in BIPs][hacspec chat]: a discussion
      about using the [hacspec][] specification language in BIPs to

>       provide specifications which are provably correct.  See also the

Reviewer 5: ```suggestion
      provide specifications that are provably correct.  See also the
```

      [transcript][hacspec preso] for a related talk during the TabConf.

    - [Package and v3 transaction relay][package relay chat]: the
      transcript of a presentation about proposals to enable [package
      transaction relay][topic package relay] and use new transaction
      relay rules to eliminate [pinning attacks][topic transaction
      pinning] in certain cases.

    - [Stratum v2][stratum v2 chat]: a discussion that started with the

>       announcement of a new open source project implementing the Stratum

Reviewer 5: ```suggestion
      announcement of a new open-source project implementing the Stratum
```
Perhaps this should be added to the [compound words list](https://github.com/bitcoinops/bitcoinops.github.io/blob/master/STYLE.md#compound-words).

  with a [post][sanders ephemeral] to the Bitcoin-Dev mailing containing
  a proposal for a new type of [anchor output][topic anchor outputs].  A
  v3 transaction could pay zero fees but contain an output paying the
  script `OP_TRUE`, allowing anyone to spend it under the consensus
  rules in a child transaction.  The zero-fee parent transaction would
  only be accepted if it was part of a transaction package which also
  contained the child transaction spending the OP_TRUE output.

    Described advantages of this proposal is that it eliminates the need
    to use one-block relative timelocks (called `1 OP_CSV` after the
    code used to enable them) to prevent [transaction pinning][topic
    transaction pinning] and allows anyone to fee bump the parent
    transaction (similar to an earlier [fee sponsorship][topic fee
    sponsorship] proposal).

    Jeremy Rubin [replied][rubin ephemeral] in support of the proposal
    but noted that it doesn't work for any contract that can't use v3
    transactions.  Several other developers also discussed the concept,
    all of them seeming to find it appealing as of this writing.

> ## Selected Q&A from Bitcoin Stack Exchange

Reviewer 5: just FYI, I verified all of the SE question numbers in what follows.

{% assign bse = "https://bitcoin.stackexchange.com/a/" %}

- [Why would someone use a 1-of-1 multisig?]({{bse}}115443)
  Vojtěch Strnad asks why someone would choose to use 1-of-1 multisig over
  P2WPKH given P2WPKH is cheaper and has a larger anonymity set. Murch lists a
  variety of resources showing at least one entity spending millions of 1-of-1
  UTXOs over the years, although the motivations remain unclear.

- [Why would a transaction have a locktime in the year 1987?]({{bse}}115549)
  1440000bytes points to a comment from Christian Decker referencing [a section][bolt 3 commitment]
  from the BOLT 3 Lightning spec that allocates the locktime field as "upper 8
  bits are 0x20, lower 24 bits are the lower 24 bits of the obscured commitment
  transaction number".

- [What is the size limit on the UTXO set, if any?]({{bse}}115439)
  Pieter Wuille answers that there is no consensus limit on the UTXO set size and that the
  rate of growth of the UTXO set is bounded by the block size which limits the
  number of UTXOs that can be created in a given block. In a [related answer][se

>   murch uxto calcs], Murch estimates that it would take about 11 years to create

Reviewer 5: uxto?
